src/base/AsyncJobCalls.h

```diff
@@ -176,9 +176,18 @@ JobDialer<Job>::dial(AsyncCall &call)
     } catch (const std::exception &e) {
         debugs(call.debugSection, 3,
                call.name << " threw exception: " << e.what());
+        if (!job) {
+            debugs(call.debugSection, 3, "ERROR: Squid BUG: Job invalidated during " <<
+                   call.name << " that threw exception: " << e.what());
+            return; // see also: bug 4981, commit e3b6f15, and XXX in Http::Stream class description
+        }
         job->callException(e);
     }
 
+    if (!job) {
+        debugs(call.debugSection, 3, "ERROR: Squid BUG: Job invalidated during " << call.name);
```

Comment:
> Amos: Are you sure that DBG_CRITICAL is the right level for these?

These checks are essentially failed assertions. Assertion failures should be reported at the highest level because they indicate that Squid may be violating all sorts of guarantees and invariants going forward.

I will restore DBG_CRITICAL (instead of the current level 3) after posting this review, but I am not going to block this PR if it ends up using DBG_IMPORTANT instead.

> If there is nothing the admin can do to avoid it perhapse level-1 would be better.

IIRC, we do not have consensus about the (difference in) meaning of debugging levels 0 and 1. Until that consensus is established, "better" is essentially undefined or, to be more precise, has many conflicting definitions. I am happy to cooperate towards establishing that consensus, but I assume that you do not want to use _this_ PR to do so.

FWIW, I disagree with all of the following statements, listed here  _without_ an implication that you are making or implying these statements:

* Level-0 messages are for things admins can prevent
* Level-1 messages are for things admins cannot prevent
* Assertions (and equivalent) are things admins cannot prevent
* Concerns about debugging noise is what determines whether level 0 or level 1 should be used.

... where "prevent" is short for "can do something to avoid in the future".

The correct way to control debugging noise[^1] among level-0/1 cache.log messages is [cache_log_message](http://www.squid-cache.org/Doc/config/cache_log_message/).

[^1]: as defined by the admin